File: scala_api/src/main/scala/application/dataAnalysis/ml/classification/Classificators.scala
package application.dataAnalysis.ml.classification

import application.dataAnalysis.Context
import application.dataAnalysis.interfaces.{ModelCreateProcess, ModelPrepare}
import application.dataAnalysis.ml.ClassificatorsAndRegressors
import application.models.Hyperparameters
import application.models.SummaryOfMerics.{ClassificationSummaryResult, SummaryResult}
import org.apache.spark.ml.PipelineStage
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.feature._
import org.apache.spark.ml.util.MLWritable
import org.apache.spark.sql.{DataFrame, Dataset, Row}

case class Classificators(
 override val data: DataFrame,
 override val train: Dataset[Row],
 override val test: Dataset[Row],
) extends ClassificatorsAndRegressors(data, train, test) with ModelCreateProcess {

  override def makeFitAndTransform(classificationModel: PipelineStage with MLWritable): DataFrame = {
    pipeline.setStages(Array(labelIndexer, featureIndexer, classificationModel, labelConverter))
    val model = pipeline.fit(train)
    val predictions: DataFrame = model.transform(test)
    if (Context.saveModels)
      safeResources(model)
    predictions
  }

  override def chooseAndCreateModel(hyperparameters: Hyperparameters, methodName: String): ModelPrepare = {
    methodName match {
      case "Logistic regression" => hyperparameters.logisticRegressionModel.getOrElse(LogisticRegressionPattern())
      case "Decision tree classifier" => hyperparameters.decisionTreeClassModel.getOrElse(DecisionTreeClassPattern())
      case "Random forest classifier" => hyperparameters.randomForrestClassModel.getOrElse(RandomForrestClassPattern())
      case "Naive Bayes" => hyperparameters.naiveBayesModel.getOrElse(LogisticRegressionPattern())
      case _ => hyperparameters.logisticRegressionModel.getOrElse(LogisticRegressionPattern())
    }
  }

  override def calculateMetrics(predictions: DataFrame): SummaryResult = {
    val evaluator = new MulticlassClassificationEvaluator()
      .setLabelCol("indexedLabel")
      .setPredictionCol("prediction")
    val accuracy = evaluator.setMetricName("accuracy").evaluate(predictions)
    val precision = evaluator.setMetricName("weightedPrecision").evaluate(predictions)
    val f1 = evaluator.setMetricName("f1").evaluate(predictions)
    val weightedRecall = evaluator.setMetricName("weightedRecall").evaluate(predictions)
    val hammingLoss = evaluator.setMetricName("hammingLoss").evaluate(predictions)
    val metrics = ClassificationSummaryResult(accuracy, 1.0 - accuracy, precision, f1, weightedRecall, hammingLoss)
    SummaryResult(classificationMetrics = Some(metrics))
  }

  private def labelIndexer: StringIndexerModel = {
    new StringIndexer()
      .setInputCol("label")
      .setOutputCol("indexedLabel")
      .fit(data)
  }

  private def labelConverter: IndexToString = {
    new IndexToString()
      .setInputCol("prediction")
      .setOutputCol("predictedLabel")
      .setLabels(labelIndexer.labelsArray(0))
  }

}

object Classificators {
  val indexOfFeatureIndexer: Int = 1
  val predictionColumn = "predictedLabel"
}

File: python_api/SparkVersion/ml/classification/Classificators.py
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml.feature import VectorIndexer, IndexToString, StringIndexer

from Context import getPipelineModelFilePath
from Models.ClassificationSummaryResult import ClassificationSummaryResult
from Models.SummaryResult import SummaryResult
from SparkVersion.ml.classification.DecisionTreeClassificationModel import DecisionTreeClassificationModel
from SparkVersion.ml.classification.LogisticRegressionModel import LogisticRegressionModel
from SparkVersion.ml.classification.NaiveBayesClassificationModel import NaiveBayesClassificationModel
from SparkVersion.ml.classification.RandomForrestClassificationModel import RandomForrestClassificationModel


class Classificators:
    def __init__(self, data, train, test, maxCategories=4):
        self.data = data
        self.train = train
        self.test = test
        self.maxCategories = maxCategories

    @staticmethod
    def chooseMethod(method, hyperparameters):
        if method == "Logistic regression":
            return LogisticRegressionModel(hyperparameters).prepareModel()
        elif method == "Decision tree classifier":
            return DecisionTreeClassificationModel(hyperparameters).prepareModel()
        elif method == "Random forest classifier":
            return RandomForrestClassificationModel(hyperparameters).prepareModel()
        else:
            return NaiveBayesClassificationModel(hyperparameters).prepareModel()

    def fitAndTransform(self, regressionModel):
        pipeline = Pipeline(stages=[self.labelIndexer(), self.featureIndexer(), regressionModel, self.labelConverter()])
        model = pipeline.fit(self.train)
        predictions = model.transform(self.test)
        model.write().overwrite().save(getPipelineModelFilePath())
        return predictions

    @staticmethod
    def calculateMetrics(predictions):
        evaluator = MulticlassClassificationEvaluator(
            labelCol="indexedLabel",
            predictionCol="prediction"
        )
        accuracy = evaluator.setMetricName("accuracy").evaluate(predictions)
        precision = evaluator.setMetricName("weightedPrecision").evaluate(predictions)
        f1 = evaluator.setMetricName("f1").evaluate(predictions)
        weightedRecall = evaluator.setMetricName("weightedRecall").evaluate(predictions)
        hammingLoss = evaluator.setMetricName("hammingLoss").evaluate(predictions)
        metrics = ClassificationSummaryResult(accuracy, 1.0 - accuracy, precision, f1, weightedRecall, hammingLoss)
        return SummaryResult(classificationMetrics=metrics)

    def labelIndexer(self):
        return StringIndexer(
            inputCol="label",
            outputCol="indexedLabel"
        ).fit(self.data)

    def featureIndexer(self):
        return VectorIndexer(
            inputCol="features",
            outputCol="indexedFeatures",
            maxCategories=self.maxCategories
        ).fit(self.data)

    def labelConverter(self):
        return IndexToString(
            inputCol="prediction",
            outputCol="predictedLabel",
            labels=self.labelIndexer().labels
        )

    @staticmethod
    def indexOfFeatureIndexer():
        return 1

    @staticmethod
    def predictionColumn():
        return "predictedLabel"
